Background process runner

ankitrox opened the pull request:
## Summary

Create a background process runner class. This is mainly responsible for:

1. Handling the background job call via ajax action.
2. Validating the parameters.
3. Running the process.
4. Watching the process for time and memory exhaustion.
5. Terminating the process.
6. Error handling.

<!-- Please reference the issue this PR fixes. If this PR does not fix the entire issue, change this to Addresses #... instead. -->
Fixes #482 #493 #494 

## Relevant technical choices

<!-- Please describe your changes. -->

## Checklist

- [x] PR has either `[Focus]` or `Infrastructure` label.
- [x] PR has a `[Type]` label.
- [x] PR has a milestone or the `no milestone` label.

Commit:
Background process runner class

Commit:
Memory check for process

Commit:
Fix: Action name for non-logged in user

Commit:
Tests: constant and hook

Commit:
Code commenting

Commit:
Fix: phpcs

Commit:
Tests: handle request

Commit:
Merge branch 'experiment/background-job-class' into experiment/bg-process-runner

Commit:
Change job status when ran successfully

To-do: Add partial status in job class

Commit:
Merge branch 'experiment/background-job-class' into experiment/bg-process-runner

Commit:
Merge branch 'experiment/background-job-class' into experiment/bg-process-runner

Commit:
Set partial status for successful batch

Commit:
Merge branch 'experiment/background-job-class' into experiment/bg-process-runner

Commit:
Fix: tests

Commit:
Update method docs

Commit:
Merge branch 'experiment/background-job-class' into experiment/bg-process-runner

Commit:
Fix: tests

Commit:
Method to call next batch of jobs

Commit:
Merge branch 'experiment/background-job-class' into experiment/bg-process-runner

jjgrainger commented on modules/images/regenerate-existing-images/background-process/class-perflab-background-process.php:
```diff
@@ -0,0 +1,236 @@
+	 *
+	 * @var string Background process action name.
+	 */
+	const BG_PROCESS_ACTION = 'background_process_handle_request';
```
@ankitrox we should prefix this action name with `perflab_`
```suggestion
	const BG_PROCESS_ACTION = 'perflab_background_process_handle_request';
```

jjgrainger commented on modules/images/regenerate-existing-images/background-process/class-perflab-background-process.php:
```diff
@@ -0,0 +1,236 @@
+			$max_execution_time = ( ! empty( $time ) && ( $time > $min_execution_time ) ) ? $time - 10 : $min_execution_time;
+		}
+
+		return ( $current_time >= ( $run_start_time + $max_execution_time ) );
```
@ankitrox we should also use `apply_filters` on the value returned here as we have with the memory limit. The filter should be `perflab_background_process_time_exceeded`
```suggestion
		$time_exceeded = ( $current_time >= ( $run_start_time + $max_execution_time ) );
		return apply_filters( 'perflab_background_process_time_exceeded', $time_exceeded );
```

jjgrainger commented on modules/images/regenerate-existing-images/background-process/class-perflab-background-process.php:
```diff
@@ -0,0 +1,236 @@
+	 * @return void
+	 */
+	public function handle_request() {
+		try {
```
@ankitrox I feel like this method and try block does too much? I think we can split some of the logic between the `handle_request()` and `run()` methods. I think this because we have exceptions thrown that don't get caught as the job is not available.

The way I see it, `handle_request()` should verify and determine the request is valid. If not, the method should return early, or maybe even return a [`wp_send_json_error`](https://developer.wordpress.org/reference/functions/wp_send_json_error/) error message? If the request is valid it then calls the `run()` method,

Then the run method is responsible for running the job and I think this could contain everything from the `$this->job->lock()` line down.

jjgrainger commented on modules/images/regenerate-existing-images/background-process/class-perflab-background-process.php:
```diff
@@ -0,0 +1,236 @@
+			 * @param array $data Job data.
+			 */
+			do_action( 'perflab_job_' . $this->job->get_name(), $this->job->get_data() );
+		} while ( ! $this->memory_exceeded() && ! $this->time_exceeded() );
```
@ankitrox we should also check if the job is marked as complete here too.

```suggestion
		} while ( ! $this->memory_exceeded() && ! $this->time_exceeded() && ! $this->job->is_completed() );
```

jjgrainger commented on modules/images/regenerate-existing-images/background-process/class-perflab-background-process.php:
```diff
@@ -0,0 +1,236 @@
+		 * Do not call the background process from within the script if the
+		 * real cron has been setup to do so.
+		 */
+		if ( defined( 'ENABLE_BG_PROCESS_CRON' ) ) {
```
@ankitrox I think the configuration we have defined so far is to disable this background process entirely with a `DISABLE_BACKGROUND_PROCESS` constant. I don't believe we have the infrastructure defined to have the WP Cron handle the additional batches.

I think we should remove this here and consider where the `DISABLE_BACKGROUND_PROCESS` check would need to be made.

ankitrox replied:
We will also have this constant to let the real cron job call background process. In such cases, script would not call next batch of the job from within itself.
This is discussed with @felixarntz in alternate approaches to improve background process workflow.

jjgrainger replied:
Thanks @ankitrox I'm not sure how this works in the grand scheme of things as we haven't defined or built the part that handles subsequent batches. I think we can approach this change in 1 of 2 ways.

1. Keep the constant as part of this PR, but make a change as this is used twice in the code, once here and again inside the `perflab_dispatch_background_process_request`. I think we should aim to only have to check this constant once. 

2. Remove the constant from the PR and we can add it as part of a new PR and stream of work to introduce this functionality.

I think in both cases we should also create a new GitHub issue with acceptance criteria to outline the work.

cc @felixarntz

jjgrainger reviewed (changes requested):
@ankitrox Looking good so far. I've left a couple of comments to address.

Additionally, there are a couple of other things. One is the failing tests on this PR the second is the inclusion of the Job class which is also in #507 and should not be present in this PR.

Commit:
Fix failing test

Commit:
Merge branch 'experiment/background-job-class' into experiment/bg-process-runner

Commit:
Execute job till completed

Commit:
Fix: while loop condition

ankitrox commented:
> @ankitrox Looking good so far. I've left a couple of comments to address.
> 
> Additionally, there are a couple of other things. One is the failing tests on this PR the second is the inclusion of the Job class which is also in #507 and should not be present in this PR.

@jjgrainger Let's aim to get [background class PR](https://github.com/WordPress/performance/pull/507) merge first. As process runner is using job instance, I have merged it in this PR. Once background job class is merged, we can get this PR ready for review to others.

Commit:
Merge branch 'experiment/background-job-class' into experiment/bg-process-runner

Commit:
Adapt job class changes

Fix: tests

Commit:
Refactor next batch method

Commit:
Refactor dispatch method

Commit:
Status check cron

Commit:
Resolve: Background job class conflicts

Commit:
Fix typo

Commit:
Move dispatch function to helper

Commit:
Merge branch 'experiment/background-job-class' into experiment/bg-process-runner

Commit:
Use get_identifier to trigger custom action

Commit:
Docblock updates

Commit:
Merge branch 'experiment/background-job-class' into experiment/bg-process-runner

Commit:
Merge branch 'experiment/background-job-class' into experiment/bg-process-runner

Commit:
Fix: method name for job action

Commit:
Merge branch 'experiment/background-job-class' into experiment/bg-process-runner

Commit:
Docblock updates

Filter to update expiry days for job cleanup

Commit:
Merge branch 'experiment/background-job-class' into experiment/bg-process-runner

Commit:
Merge branch 'feature/regenerate-existing-images' into experiment/bg-process-runner

jjgrainger commented:
@ankitrox I think this PR is becoming too large and tackling too many parts of the background process.

For starters, I would argue the work for status cron #484 can be moved out of the background process runner and into its own PR/file/class. I think this should definitely be moved out of the PR and into a new one focused on fixing that specific issue.

Additionally, this PR seems to also fix #480 #481 and even #490 #483, which were already being addressed in other PR's. We want to keep these PR's smaller to make the review process more focused and less complex.

Can you update this PR to focus on the acceptance criteria in #482 #493 #494.

Let me know if you have any questions.

Commit:
Remove irrelevant code

Commit:
Helper method to get job instance

Commit:
Tests: perflab_get_background_job

ankitrox commented:
> @ankitrox I think this PR is becoming too large and tackling too many parts of the background process.
> 
> For starters, I would argue the work for status cron #484 can be moved out of the background process runner and into its own PR/file/class. I think this should definitely be moved out of the PR and into a new one focused on fixing that specific issue.
> 
> Additionally, this PR seems to also fix #480 #481 and even #490 #483, which were already being addressed in other PR's. We want to keep these PR's smaller to make the review process more focused and less complex.
> 
> Can you update this PR to focus on the acceptance criteria in #482 #493 #494.
> 
> Let me know if you have any questions.

Thanks @jjgrainger 
I have removed the cron related code from PR. I will create a new PR for that one.
Also, updated the dependent issue in PR description.

Commit:
Tests: get background job

jjgrainger commented on modules/images/regenerate-existing-images/background-process/class-perflab-background-process.php:
```diff
@@ -0,0 +1,247 @@
+	public function __construct() {
+		// Handle job execution request.
+		add_action( 'wp_ajax_' . Perflab_Background_Process::BG_PROCESS_ACTION, array( $this, 'handle_request' ) );
+		add_action( 'wp_ajax_nopriv_' . Perflab_Background_Process::BG_PROCESS_ACTION, array( $this, 'handle_request' ) );
```
@ankitrox we should remove this action as we do not want unauthenticated requests triggering the background process. We should stick with the `wp_ajax_perflab_background_process_handle_request` action only.

jjgrainger commented on modules/images/regenerate-existing-images/helper.php:
```diff
@@ -71,3 +71,66 @@ function perflab_delete_background_job( $job_id ) {
+ *
+ * @param int $job_id Job ID.
+ */
+function perflab_dispatch_background_process_request( $job_id ) {
```
@ankitrox based on the acceptance criteria in #480 this function should be named `perflab_start_background_job`

jjgrainger commented on modules/images/regenerate-existing-images/background-process/class-perflab-background-process.php:
```diff
@@ -0,0 +1,247 @@
+			return;
+		}
+
+		perflab_dispatch_background_process_request( $job_id );
```
@ankitrox we are handling next batch requests differently to the start request. See #490 for the AC's here. The function name that should be called is `perflab_background_process_next_batch`.

ankitrox replied:
There's only a difference of `$key` in `start_job` and `next_batch` functions. I think we can somehow combine/compose these functions to avoid rewriting entire thing twice.

jjgrainger commented on modules/images/regenerate-existing-images/background-process/class-perflab-background-process.php:
```diff
@@ -0,0 +1,247 @@
+	 *
+	 * @throws Exception Invalid request for background process.
+	 */
+	public function handle_request() {
```
@ankitrox I think this method does too much and should only be responsible for handling the HTTP request and calling the `run()` method. I would make the following changes:

- Remove the try/catch block and exceptions - as we're handling the HTTP request we can return early with a error response. This could be a combination of using `WP_Error` instead of `Exception` and returning it with [`wp_send_json_error`](https://developer.wordpress.org/reference/functions/wp_send_json_error/)
- The final part of the `handle_request()` method should be the `$this->run()` call.
- The `unlock()` and `next_batch()` lines should be moved to the end of the `run()` method. Reasoning here is that this logic is part of the job running processes. If we were to trigger running a job via a different method (WP_Cron) calling `$this->run()` would still work as expected.
- If you wanted you can use a try/catch block inside the `run()` method to catch any system errors while jobs are running and handle those properly.

Let me know if this makes sense, happy to discuss on a call

ankitrox replied:
Made the proposed changes.
I think adding `try catch` to run method would make more sense when we will implement error handling.

jjgrainger reviewed (changes requested):
Thanks @ankitrox left some comments but looking good so far

Commit:
Refactor handle_request method

ankitrox commented:
> Thanks @ankitrox I'm not sure how this works in the grand scheme of things as we haven't defined or built the part that handles subsequent batches. I think we can approach this change in 1 of 2 ways.
> 
> Keep the constant as part of this PR, but make a change as this is used twice in the code, once here and again inside the perflab_dispatch_background_process_request. I think we should aim to only have to check this constant once.
> 
>  Remove the constant from the PR and we can add it as part of a new PR and stream of work to introduce this functionality.
> 
>  I think in both cases we should also create a new GitHub issue with acceptance criteria to outline the work.
> 
>  cc @felixarntz
> 

@jjgrainger  Regarding this comment, I think option #2 is better; to handle this in separate issue and PR.

Commit:
Refactor request triggering function

Commit:
Tests: start process and next batch process

Commit:
Tests: helper functions for batch call

adamsilverstein commented:
@ankitrox looks like this PR has gone stale, can we go ahead and close it?

adamsilverstein reviewed (commented):
is this still needed?

westonruter commented:
@adamsilverstein looking back at old PRs from this time period, it seems the main push was to add image regeneration as a background process which this PR would serve as infrastructure for. It seems this topic hasn't been on the radar for quite some time, so potentially several such PRs should be closed. Not sure why that got deprioritized.

swissspidy commented:
Closing for now to clean up the repo since this plugin/module is not a priority right now.
